tools/codegen/api/cpp.py

```diff
@@ -197,21 +195,17 @@ def default_expr(d: str, t: Type) -> str:
     return JIT_TO_CPP_DEFAULT.get(d, d)
 
 # Convert an argument into its C++ API form
-def argument(a: Union[Argument, TensorOptionsArguments, ThisArgument]) -> CppArgument:
+
+def argument_not_this(
```

Comment:
Hmm. Still thinking about the overall comment, but avoiding `assert_never` is not something you should specifically go for, all it is, is a way of making the typechecker check that the isinstance cases (aka the pattern match) is exhaustive. It's not an antipattern to use it and we should use it as much as possible.

Reply:
Re the broader comment, I don't think I really care too much either way. The code is written in its current form because I was trying to minimize the number of top level functions I needed, and so if I write the function to take a Union I only need to write one function (versus two, and also needing to do pattern matching at the caller.)

The beautiful thing about using mypy is that there is almost no precision difference between the two different schemes: the only benefit to the more factored representation is you get to skip an isinstance check when you statically know you only have an Argument or TensorOptionsArgument, but we don't really care about this kind of microoptimization here.

Reply:
Oh of course, it's not about avoiding assert_never per se, more about avoiding partial functions even at the cost of a marginally higher total number of functions (though not at the cost of redundant code). The thought was that since all callers were total over the supertype, cracking this helper avoid the partial function and hence the assert_never. (But I can't say I verified that this factoring would definitely work.)

Reply:
> avoiding partial functions

There isn't a partial function here! `assert_never` doesn't mean "this function is partial", it means, "I am doing a case match on a Union and this case match is exhaustive".